### PR TITLE
Complex support for basic arithmetic | feat(torchlib)

### PR DESCRIPTION
**Pull request by justinchuby**
Implement support for complex inputs for `+`, `-`, `*`, `/`.

Updated aten_abs to use Slice instead of Gather for runtime performance.

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -167,6 +167,13 @@ def aten_add(self: TReal, other: TReal, alpha: float = 1.0) -> TReal:
     return op.Add(self, other)
 
 
+@torch_op(("aten::add", "aten::add.Tensor", "_operator::add"), trace_only=True, complex=True)
+def aten_add_complex(self: TReal, other: TReal, alpha: float = 1.0) -> TReal:
+    """add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor"""
+
+    return aten_add(self, other, alpha=alpha)
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/ops/core.py:174]: is computing logic same for complex with `alpha`?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 2]
```diff
@@ -2630,6 +2637,40 @@ def aten_div(self: TFloat, other: TFloat) -> TFloat:
...
+    self_imag = op.Slice(self, [1], [2], axes=[-1])
+    other_real = op.Slice(other, [0], [1], axes=[-1])
+    other_imag = op.Slice(other, [1], [2], axes=[-1])
+
+    # Complex division
+    # (a + bi) / (c + di) = (ac + bd) / (c^2 + d^2) + (bc - ad) / (c^2 + d^2)i
+    # https://mathworld.wolfram.com/ComplexDivision.html
+    real = op.Div(
+        op.Add(op.Mul(self_real, other_real), op.Mul(self_imag, other_imag)),
+        op.Add(op.Mul(other_real, other_real), op.Mul(other_imag, other_imag)),
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/ops/core.py:2664]: nit: `op.Add(op.Mul(other_real, other_real), op.Mul(other_imag, other_imag)),` this part is repeated

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:174]: I think so, because alpha is just a scaling factor.

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:2664]: this is taking the power. did you see repeated code?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 3]
```diff
@@ -135,13 +135,13 @@ def aten_abs(self: TRealOrUInt8) -> TRealOrUInt8:
...
-    self_real = op.Gather(self, 0, axis=-1)
+    self_real = op.Slice(self, [0], [1], axes=[-1])
     # self_imag = self[..., 1]
-    self_imag = op.Gather(self, 1, axis=-1)
+    self_imag = op.Slice(self, [1], [2], axes=[-1])
     real_pow = op.Pow(self_real, 2)
     imag_pow = op.Pow(self_imag, 2)
     real_plus_imag = op.Add(real_pow, imag_pow)
-    return op.Sqrt(real_plus_imag)
+    return op.Squeeze(op.Sqrt(real_plus_imag), axes=[-1])
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/ops/core.py:144]: Could you explain why these changes for `aten_abs_complex`?

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:144]: Updated PR description: Updated aten_abs to use Slice instead of Gather for runtime performance.

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 2 above)
Comment (BowenBao) [reply onnxscript/function_libs/torch_lib/ops/core.py:2664]: repeated for real and imag?

**Review by BowenBao — APPROVED** [review]

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:2664]: I see sorry. Let me update

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py]: Done

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 4]
```diff
@@ -5319,6 +5358,29 @@ def aten_mul_bool(self: BOOL, other: BOOL) -> BOOL:
     return op.And(self, other)
 
 
+@torch_op(("aten::mul", "aten::mul.Tensor", "_operator::mul"))
```
Comment (titaiwangms) [inline onnxscript/function_libs/torch_lib/ops/core.py:5361]: This is missing complex=True? Making regression on CI failing https://github.com/pytorch/pytorch/pull/112758

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 5]
```diff
@@ -2630,6 +2637,39 @@ def aten_div(self: TFloat, other: TFloat) -> TFloat:
...
 
+@torch_op(
+    (
+        "aten::div",
+        "aten::div.Tensor",
+        "aten::div.Scalar",
+        "aten::divide",
+        "aten::true_divide",
+        "_operator::truediv",
+    )
```
Comment (titaiwangms) [inline onnxscript/function_libs/torch_lib/ops/core.py:2648]: Also missing

Comment (titaiwangms) [reply onnxscript/function_libs/torch_lib/ops/core.py:2648]: https://github.com/microsoft/onnxscript/pull/1146
